File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    
    private Map<String, Integer> combinationMap;

    public String[] solution(String[] orders, int[] course) {
        List<String> answer = new ArrayList<>();

        for (int size : course) {
            combinationMap = new HashMap<>();
            int maxCount = 0;

            for (String order : orders) {
                char[] chars = order.toCharArray();
                Arrays.sort(chars);
                generateCombinations(chars, 0, size, new StringBuilder());
            }

            for (int count : combinationMap.values()) {
                maxCount = Math.max(maxCount, count);
            }

            if (maxCount < 2) {
                continue;
            }

            for (Map.Entry<String, Integer> entry : combinationMap.entrySet()) {
                if (entry.getValue() == maxCount) {
                    answer.add(entry.getKey());
                }
            }
        }

        Collections.sort(answer);
        return answer.toArray(new String[0]);
    }

    private void generateCombinations(char[] chars, int index, int targetSize, StringBuilder current) {
        if (current.length() == targetSize) {
            String key = current.toString();
            combinationMap.put(key, combinationMap.getOrDefault(key, 0) + 1);
            return;
        }

        if (index == chars.length) {
            return;
        }

        current.append(chars[index]);
        generateCombinations(chars, index + 1, targetSize, current);
        current.deleteCharAt(current.length() - 1);

        generateCombinations(chars, index + 1, targetSize, current);
    }
}

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
from collections import defaultdict, Counter
from itertools import combinations


def solution(orders, course):
    answer = []
    counter = defaultdict(int)
    
    for order in orders:
        for c in course:
            for comb in combinations(order, c):
                counter[''.join(sorted(comb))] += 1
    
    menus = defaultdict(list)
    for name, count in counter.items():
        menus[len(name)].append((name, count))
    
    for menu_lst in menus.values():
        max_count = max(x[1] for x in menu_lst)
        for o, v in sorted(menu_lst, key=lambda x: x[1], reverse=True):
            if v != max_count or v == 1:
                break
            answer.append(o)
        
    return sorted(answer)